### DISCUSSION TITLE
Help building from source Linux

### DISCUSSION THREAD

**Discussion by Myrkskog** [andrei-drexler/ironwail#140] [Q&A]
Attempting to build using VSCodium as IDE. It suggests installation of intersystems language server. Despite obtaining these extensions I continue to receive a notification that intersystems is either disabled or not installed.
Trying my best to configure intersystems but am left confused. Is this necessary?

Anyone have a guide to help build Ironwail on Linux? Using Ubuntu22.04 and trying my best but it sure is a time sink. 

Thank you in advance for any help.

**Comment by andrei-drexler** [comment]
I think the easiest way to build it on Linux is from the terminal, without any IDE. You'll need to make sure `libsdl2-dev`, `libvorbis-dev` and `libmad0-dev` are already installed (apt-get install them if not). Then, navigate to where your source code is and
```bash
cd Quake
make clean
make -j
```
should produce a binary named `ironwail` in the `Quake` subdir. If you have Quake on Steam (using Proton) you should be able to run the binary directly (it should find the data files on its own), otherwise you'll need to use `-basedir` to specify the game files are.

**Reply by Myrkskog** [reply]
Thank you Andrei. I am that much closer to trying Ironwail. Unfortunately, it did not find my pak0.pak. I tried running Ironwail from terminal with -basedir pointing to /home/.steam/steam/common/steam apps/Quake with no luck. Ironwail starts and closes once trying to load demo sequence as the console retracts. I renamed PAK0.PAK to pak0.pak in event there is case sensitivity.

**Reply by andrei-drexler** [reply]
Sorry about the late reply! You're right about the case sensitivity, you either need to rename `Id1`, `PAK0.PAK` and `PAK1.PAK` to be lowercase, or set `-basedir` to the `rerelease` dir (which is what Ironwail does on Linux if it finds a Steam Quake installation). If you're seeing the game console, though, it means that data files are loaded from the specified basedir, so at that point everything should work. I can't really tell what's going on without more info - there should be some kind of error message in the terminal. If not, maybe the game is actually crashing, in which case you might be able to get a stack trace by compiling a debug build with
```bash
make clean
make DEBUG=1 -j
````
and then running it with gdb
```bash
gdb -ex=run --args ./ironwail -basedir "path/to/quake/dir"
```
and then waiting for the game to crash. When that happens, entering the `bt` command in gdb should print a stack trace. I realize that's a bit cumbersome, though, I'll see about making some actual project files for VSCode to make debugging easier in the future.

**Comment by Myrkskog** [comment]
Fantastic! Thank you for the thorough reply. I will try this out. I
appreciate your help very much. Busy for the next few days, so hang tight 👍

**Comment by daftmugi** [comment]
This may just cause some confusion, but I thought I'd add how I build Ironwail on Ubuntu 22.04 in case it helps. Sometimes starting over and trying again can help. If not, running `gdb` like Andrei stated above is the next step if the game is crashing. Also, if you can `./ironwail > output.txt` and share the resulting `output.txt` or some of it, that may also be useful in diagnosing what's going on.

1. Install packages (I may have a few extra packages here, since I build multiple games.)
   ```
   sudo apt install udev build-essential make gcc libsdl2-dev libvorbis-dev libmad0-dev libx11-xcb-dev libcurl4-openssl-dev
   ```
   Note: I found installing `udev` first was important for dependency resolution.
   Note: `libcurl4-openssl-dev` (or another `libcurl4-*-dev` variant is needed as of 7ab6d8a392797344fb35e8ec74a36a8c6b47ce4e)

2. Build Ironwail
   ```
   cd src/ironwail/Quake
   make clean
   make -j
   ```

3. Copy `ironwail` executable and `ironwail.pak` to its game directory

   ```
   cp src/ironwail/Quake/ironwail      games/ironwail/ironwail
   cp src/ironwail/Quake/ironwail.pak  games/ironwail/ironwail.pak
   ```

4. The resulting file tree should look like:

   ```
   ./games/ironwail
   ├── id1
   │   ├── music
   │   ├── pak0.pak
   │   └── pak1.pak
   ├── ironwail
   └── ironwail.pak
   ```
   Note: Lowercase `id1`, `pak0.pak`, `pak1.pak`. The `music` directory contains music files: `track02.wav`, `track03.wav` ... `track11.wav` (can be `wav`, `ogg`, or another format).

5. It may be a good idea to delete any `*.cfg` files you have in `id1/`, such as `id1/config.cfg` and `id1/autoexec.cfg`, while you are troubleshooting.

6. You can check to make sure your `pak0.pak` and `pak1.pak` files are right by checking their SHA256-SUM:
   ```
   $ sha256sum *pak
   35a9c55e5e5a284a159ad2a62e0e8def23d829561fe2f54eb402dbc0a9a946af  pak0.pak
   94e355836ec42bc464e4cbe794cfb7b5163c6efa1bcc575622bb36475bf1cf30  pak1.pak
   ```

7. Run ironwail
   ```
   ./ironwail
   ```

**Reply by Myrkskog** [reply]
Thank you both for the help. I followed both of your instructions. Turns out it compiled fine. The trouble was not using dedicated graphics. My system was defaulting to Intel 4400 integrated graphics. Now it's running fine and super fast I might add. I am very pleased with how Ironwail runs on my laptop. Excellent stuff 👍

**Comment by daftmugi** [comment]
Note: As of 7ab6d8a392797344fb35e8ec74a36a8c6b47ce4e, `libcurl4-openssl-dev` or another `libcurl4-*-dev` variant is needed to compile successfully.

**Comment by stemid** [comment]
I found this thread so I thought I'd add a short guide on how I built ironwail in 2023-11-28 on Fedora 38.

```
sudo dnf install SDL2-devel libvorbis-devel libmad-devel pkgconf libcurl-devel -y
cd Quake
make clean && make -j
./ironwail -basedir ~/.var/app/com.valvesoftware.Steam/.local/share/Steam/steamapps/common/Quake/rerelease
```

I use the Steam flatpak, hence the basedir path. If you use the Steam RPM package I'm sure ironwail will find the files on its own.

**Comment by alfasketer** [comment]
Hi, similar problem here. I built Ironwail on my Ubuntu 24.04, and after I launch it, it immediately freezes. It exits with SIGABRT and I can't figure out why does it abort. My GPU is Intel HD 4400, with Mesa 24 and OpenGL version 4.6.
For context Ironwail worked wonders on the same machine when it was loaded with Windows, but I needed to switch OSes.
I will leave the full debug log bellow.

Thanks in advance. 

```
$:~/repos/ironwail/Quake$ gdb -ex=run --args ./ironwail -basedir "/home/alfa/quake/"
GNU gdb (Ubuntu 15.0.50.20240403-0ubuntu1) 15.0.50.20240403-git
Copyright (C) 2024 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./ironwail...
Starting program: /home/alfa/repos/ironwail/Quake/ironwail -basedir /home/alfa/quake/

This GDB supports auto-downloading debuginfo from the following URLs:
  <https://debuginfod.ubuntu.com>
Enable debuginfod for this session? (y or [n]) y
Debuginfod has been enabled.
To make this setting permanent, add 'set debuginfod enabled on' to .gdbinit.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Command line: /home/alfa/repos/ironwail/Quake/ironwail -basedir /home/alfa/quake/
Found SDL version 2.30.0
Detected 4 CPUs.
Initializing Ironwail v0.7.0-dev
Host_Init

Using add-on server "https://kexquake.s3.amazonaws.com"
[New Thread 0x7fffdd8006c0 (LWP 44781)]
[New Thread 0x7fffdce006c0 (LWP 44782)]
Playing registered version.
Downloading separate debug info for /lib/x86_64-linux-gnu/libnss_mdns4_minimal.so.2
Downloading separate debug info for /lib/x86_64-linux-gnu/libnss_mdns4_minimal.so.2                                                                                                                                                         
[New Thread 0x7fffd7e006c0 (LWP 44784)]                                                                                                                                                                                                     
Using renderer/network isolation.
Console initialized.
UDP Initialized
Server using protocol 999 (RMQ)
Exe: 09:58:25 Oct  1 2024 (Linux 64-bit)
384.0 megabyte heap

Video initialization
[Thread 0x7fffdce006c0 (LWP 44782) exited]
Downloading separate debug info for /lib/x86_64-linux-gnu/libicuuc.so.74
Downloading separate debug info for /lib/x86_64-linux-gnu/libicudata.so.74                                                                                                                                                                  
[New Thread 0x7fffdce006c0 (LWP 44785)]                                                                                                                                                                                                     
[New Thread 0x7fffc3a006c0 (LWP 44786)]
[Thread 0x7fffc3a006c0 (LWP 44786) exited]
[New Thread 0x7fffc3a006c0 (LWP 44787)]
Video mode: 1920 x 1080 24bit Z24 S8 60Hz
GL_VENDOR:   Intel
GL_RENDERER: Mesa Intel(R) HD Graphics 4400 (HSW GT2)
GL_VERSION:  4.6 (Core Profile) Mesa 24.0.9-0ubuntu0.1
FOUND: EXT_texture_filter_anisotropic
FOUND: ARB_buffer_storage
FOUND: ARB_multi_bind
FOUND: ARB_clip_control
Intel Display Adapter detected, enabling gl_clear
GL shader compiler [#1/high]: 0:34(6): warning: some implementations may not support implicit int -> uint conversions for `&' operators; consider casting explicitly for portability
GL shader compiler [#2/high]: 0:35(6): warning: some implementations may not support implicit int -> uint conversions for `&' operators; consider casting explicitly for portability
GL shader compiler [#3/high]: 0:107(24): warning: `out_fragcolor' used uninitialized
Downloading separate debug info for /lib/x86_64-linux-gnu/libudev.so.1
                                                                                                                                                                                                                                            
Sound Initialization
[New Thread 0x7fffba4006c0 (LWP 44788)]
[New Thread 0x7fffd7eff6c0 (LWP 44789)]
[New Thread 0x7fffb9a006c0 (LWP 44790)]
SDL audio spec  : 44100 Hz, 1024 samples, 2 channels
SDL audio driver: pulseaudio - Built-in Audio Analog Stereo, 65536 bytes buffer
Audio: 16 bit, stereo, 44100 Hz
[New Thread 0x7fffafe006c0 (LWP 44793)]

Language initialization
[Thread 0x7fffafe006c0 (LWP 44793) exited]
Steam library not found.
Couldn't load 'localization/loc_english.txt'
Steam library not found.
Couldn't load 'localization/loc_english.txt'

========= Quake Initialized =========

execing quake.rc
execing default.cfg
execing ironwail.cfg
Unknown command "joy_enable"
couldn't exec autoexec.cfg
3 demo(s) in loop
Playing demo from demo1.dem.

----------------------------------------

the Necropolis
Using protocol 15
[Thread 0x7fffdd8006c0 (LWP 44781) exited]

Add-on server status:
 26 add-ons available for download
  0 add-ons already installed


Thread 1 "ironwail" received signal SIGABRT, Aborted.
Download failed: Invalid argument.  Continuing without source file ./nptl/./nptl/pthread_kill.c.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:44
warning: 44     ./nptl/pthread_kill.c: No such file or directory
(gdb) continue 
Continuing.
Couldn't get registers: No such process.
(gdb) [Thread 0x7fffb9a006c0 (LWP 44790) exited]
[Thread 0x7fffd7eff6c0 (LWP 44789) exited]
[Thread 0x7fffba4006c0 (LWP 44788) exited]
[Thread 0x7fffdce006c0 (LWP 44785) exited]
[Thread 0x7fffd7e006c0 (LWP 44784) exited]
[Thread 0x7ffff5bc5340 (LWP 44778) exited]
[Thread 0x7fffc3a006c0 (LWP 44787) exited]
[New process 44778]

Program terminated with signal SIGABRT, Aborted.
The program no longer exists.
exit
```
